### PR TITLE
Mend gaps in meshes caused by trigonometric funcs.

### LINKED ISSUE #93865: Round primitive meshes contain gaps.
### Tested versions

v4.3.beta2.official [b75f0485b]

Tested version on v4.3.beta2.official [b75f0485b], but likely has been present for a long time. First I noticed it https://github.com/godotengine/godot/pull/80710#issuecomment-1690749054 (and subsequently forgot.)

### System information

Godot v4.3.beta2 - Windows 10.0.22631 - GLES3 (Compatibility) - NVIDIA GeForce RTX 4090 (NVIDIA; 31.0.15.3699) - AMD Ryzen 9 7950X 16-Core Processor (32 Threads)

### Issue description

Primitive meshes that contain round surfaces, i.e. SphereMesh, CapsuleMesh, CylinderMesh and TorusMesh, have minuscule gaps. This is due to the fact that some of the vertices at the seams do not align. The gaps are unlikely to be seen in regular use but here's an image of an unconventional shader that exaggerate them:

capsule_and_sphere

Capsule and sphere, using a shader that expands vertices based on their floating-point presentations mantissa (for visualization purposes only; do not use the shader to debug the issue, as it might be inaccurate):

```glsl
shader_type spatial;
render_mode cull_disabled, unshaded;

void vertex() {
	vec3 n = normalize(vec3(floatBitsToUint(VERTEX) & uvec3(255)) - 127.5);
	VERTEX += NORMAL * (dot(NORMAL, n) + 1.) * 0.05;
}

void fragment() {
	ALBEDO = FRONT_FACING ? vec3(0.) : vec3(1., 0., 1.);
}
```

The gaps are more of an issue when manipulating the meshes in code, e.g. when iterating unique vertices to create edge-meshes.

The cause is likely a precision issue caused by calculating the same position twice with different values. Example code from CylinderMesh:

https://github.com/godotengine/godot/blob/9db1a963beae8056cbd30d692d4160d09c10b2dc/scene/resources/3d/primitive_meshes.cpp#L1010-L1015

 (`sin(0)` and `sin(TAU)` aren't identical)

For a fix I suggest either reusing values for positions already calculated or ensuring that the values used to derive the positions are identical.

### Steps to reproduce

The unaligned vertices can be for example found using the following script:
```gdscript
extends Node

func _ready() -> void:
	print_gaps(SphereMesh.new())
	print_gaps(CapsuleMesh.new())
	print_gaps(CylinderMesh.new())
	print_gaps(TorusMesh.new())

func print_gaps(mesh: PrimitiveMesh) -> void:
	var vertex_array: PackedVector3Array = mesh.get_mesh_arrays()[Mesh.ArrayType.ARRAY_VERTEX]
	for i: int in vertex_array.size():
		var v1: Vector3 = vertex_array[i]
		for j: int in range(i+1, vertex_array.size()):
			var v2: Vector3 = vertex_array[j]
			if v1 != v2 and v1.is_equal_approx(v2):
				printt("Gap:", v1, v2)
```

### Minimal reproduction project (MRP)

Project containing both the shader used to demonstrate the gaps and the script to find them:
[primitive_meshes.zip](https://github.com/user-attachments/files/16069240/primitive_meshes.zip)

### PR DESCRIPTION
**Pull request by Lielay9**
Fixes: #93865
Supersedes: #94199

Simple, ugly solution: Ensure last iteration uses the same values as the first. Here's a cheat-sheet to help reviewing:
```
sin(τ) = 0
cos(τ) = 1
sin(π) = 0
cos(π) = -1
sin(π/2) = 1
cos(π/2) = 0
sin(0) = 0
cos(0) = 1
```

### CODE REVIEW

File: scene/resources/3d/primitive_meshes.cpp [hunk 1]
```diff
@@ -527,24 +542,33 @@ void CapsuleMesh::create_mesh_array(Array &p_arr, const float radius, const floa
...
-		v += 1.0;
-		w = sin(0.5 * Math_PI * v);
-		y = radius * cos(0.5 * Math_PI * v);
+		if (j == (rings + 1)) {
+			w = 0.0;
+			y = -radius;
+		} else {
+			w = cos(0.5 * Math_PI * v);
+			y = -sin(0.5 * Math_PI * v) * radius;
+		}
```
Comment (Lielay9) [inline scene/resources/3d/primitive_meshes.cpp:551]: Avoid unnecessary addition and subtraction.
```
sin((x+1)π/2) == cos(xπ/2)
cos((x+1)π/2) == -sin(xπ/2)
```

**Comment by Lielay9** [comment]
> Hi, I also work on CSG, would it be ok to also test CSGMesh3D too?

_Originally posted by @fire in https://github.com/godotengine/godot/issues/94199#issuecomment-2519961089_

Tested CSG-meshes! Since only the sphere had gaps, I stuffed it to this pr~

**Comment by fire** [comment]
I suspect the capsule also has the same gap.

**Comment by Lielay9** [comment]
> I suspect the capsule also has the same gap.

I don't think we have CSGCapsules...

**Comment by Lielay9** [comment]
Ah! I see! All the primitive meshes that had gaps are already included CapsuleMesh, CylinderMesh, SphereMesh, TorusMesh and TubeTrailMesh ^^/  This means there shouldn't be gaps in CSGMesh3D either! Or at least those that caused by the trig functions... but I haven't found other causes and my naive test passes for all the shapes.

**Review by fire — APPROVED** [review]
I have tested all primitive mesh shapes except those that aren't mathematically usable with CSG.

The manifold definition requires the meshes to be watertight.

Note that the pull request is insufficient; will also need the Godot Engine master as of https://github.com/godotengine/godot/commit/eb5103093c1bfd4d527ec1255d28e9bc8d3625b5 or newer

image

[csg-mend-gaps-game-project.zip](https://github.com/user-attachments/files/18030074/csg-mend-gaps-game-project.zip)

Edited:

Did not do a style review.

File: scene/resources/3d/primitive_meshes.cpp [hunk 2]
```diff
@@ -443,15 +443,25 @@ void CapsuleMesh::create_mesh_array(Array &p_arr, const float radius, const floa
...
 			u /= radial_segments;
 
-			x = -sin(u * Math_TAU);
-			z = cos(u * Math_TAU);
+			if (i == radial_segments) {
+				x = 0.0;
+				z = 1.0;
+			} else {
+				x = -sin(u * Math_TAU);
+				z = cos(u * Math_TAU);
```
Comment (AThousandShips) [inline scene/resources/3d/primitive_meshes.cpp:463]:
```suggestion
				x = -Math::sin(u * Math_TAU);
				z = Math::cos(u * Math_TAU);
```

File: scene/resources/3d/primitive_meshes.cpp [hunk 3]
```diff
@@ -492,8 +502,13 @@ void CapsuleMesh::create_mesh_array(Array &p_arr, const float radius, const floa
...
```
Comment (AThousandShips) [inline scene/resources/3d/primitive_meshes.cpp:510]:
```suggestion
				x = -Math::sin(u * Math_TAU);
				z = Math::cos(u * Math_TAU);
```

File: scene/resources/3d/primitive_meshes.cpp [hunk 4]
```diff
@@ -527,24 +542,33 @@ void CapsuleMesh::create_mesh_array(Array &p_arr, const float radius, const floa
...
 		v /= (rings + 1);
-		v += 1.0;
-		w = sin(0.5 * Math_PI * v);
-		y = radius * cos(0.5 * Math_PI * v);
+		if (j == (rings + 1)) {
+			w = 0.0;
+			y = -radius;
+		} else {
+			w = cos(0.5 * Math_PI * v);
+			y = -sin(0.5 * Math_PI * v) * radius;
```
Comment (AThousandShips) [inline scene/resources/3d/primitive_meshes.cpp:550]:
```suggestion
			w = Math::cos(0.5 * Math_PI * v);
			y = -Math::sin(0.5 * Math_PI * v) * radius;
```

File: scene/resources/3d/primitive_meshes.cpp [hunk 5]
```diff
@@ -527,24 +542,33 @@ void CapsuleMesh::create_mesh_array(Array &p_arr, const float radius, const floa
...
 			u /= radial_segments;
 
-			x = -sin(u * Math_TAU);
-			z = cos(u * Math_TAU);
+			if (i == radial_segments) {
+				x = 0.0;
+				z = 1.0;
+			} else {
+				x = -sin(u * Math_TAU);
+				z = cos(u * Math_TAU);
```
Comment (AThousandShips) [inline scene/resources/3d/primitive_meshes.cpp:562]:
```suggestion
				x = -Math::sin(u * Math_TAU);
				z = Math::cos(u * Math_TAU);
```

File: scene/resources/3d/primitive_meshes.cpp [hunk 6]
```diff
@@ -1074,8 +1098,13 @@ void CylinderMesh::create_mesh_array(Array &p_arr, float top_radius, float botto
...
 			u /= radial_segments;
 
-			x = sin(u * Math_TAU);
-			z = cos(u * Math_TAU);
+			if (i == radial_segments) {
+				x = 0.0;
+				z = 1.0;
+			} else {
+				x = sin(u * Math_TAU);
+				z = cos(u * Math_TAU);
```
Comment (AThousandShips) [inline scene/resources/3d/primitive_meshes.cpp:1106]:
```suggestion
				x = Math::sin(u * Math_TAU);
				z = Math::cos(u * Math_TAU);
```

File: scene/resources/3d/primitive_meshes.cpp [hunk 7]
```diff
@@ -1126,8 +1155,13 @@ void CylinderMesh::create_mesh_array(Array &p_arr, float top_radius, float botto
...
 			r /= radial_segments;
 
-			x = sin(r * Math_TAU);
-			z = cos(r * Math_TAU);
+			if (i == radial_segments) {
+				x = 0.0;
+				z = 1.0;
+			} else {
+				x = sin(r * Math_TAU);
+				z = cos(r * Math_TAU);
```
Comment (AThousandShips) [inline scene/resources/3d/primitive_meshes.cpp:1163]:
```suggestion
				x = Math::sin(r * Math_TAU);
				z = Math::cos(r * Math_TAU);
```

File: scene/resources/3d/primitive_meshes.cpp [hunk 8]
```diff
@@ -1168,8 +1202,13 @@ void CylinderMesh::create_mesh_array(Array &p_arr, float top_radius, float botto
...
```
Comment (AThousandShips) [inline scene/resources/3d/primitive_meshes.cpp:1210]:
```suggestion
				x = Math::sin(r * Math_TAU);
				z = Math::cos(r * Math_TAU);
```

File: scene/resources/3d/primitive_meshes.cpp [hunk 9]
```diff
@@ -1934,15 +1973,25 @@ void SphereMesh::create_mesh_array(Array &p_arr, float radius, float height, int
...
 
 		v /= (rings + 1);
-		w = sin(Math_PI * v);
-		y = scale * cos(Math_PI * v);
+		if (j == (rings + 1)) {
+			w = 0.0;
+			y = -scale;
+		} else {
+			w = sin(Math_PI * v);
+			y = scale * cos(Math_PI * v);
```
Comment (AThousandShips) [inline scene/resources/3d/primitive_meshes.cpp:1981]:
```suggestion
			w = Math::sin(Math_PI * v);
			y = scale * Math::cos(Math_PI * v);
```

File: scene/resources/3d/primitive_meshes.cpp [hunk 10]
```diff
@@ -1934,15 +1973,25 @@ void SphereMesh::create_mesh_array(Array &p_arr, float radius, float height, int
...
 			u /= radial_segments;
 
-			x = sin(u * Math_TAU);
-			z = cos(u * Math_TAU);
+			if (i == radial_segments) {
+				x = 0.0;
+				z = 1.0;
+			} else {
+				x = sin(u * Math_TAU);
+				z = cos(u * Math_TAU);
```
Comment (AThousandShips) [inline scene/resources/3d/primitive_meshes.cpp:1993]:
```suggestion
				x = Math::sin(u * Math_TAU);
				z = Math::cos(u * Math_TAU);
```

File: scene/resources/3d/primitive_meshes.cpp [hunk 11]
```diff
@@ -2434,8 +2483,12 @@ void TubeTrailMesh::_create_mesh_array(Array &p_arr) const {
 			if (curve.is_valid() && curve->get_point_count() > 0) {
 				r *= curve->sample_baked(v);
 			}
-			float x = sin(u * Math_TAU);
-			float z = cos(u * Math_TAU);
+			float x = 0.0;
+			float z = 1.0;
+			if (i < radial_steps) {
+				x = sin(u * Math_TAU);
+				z = cos(u * Math_TAU);
```
Comment (AThousandShips) [inline scene/resources/3d/primitive_meshes.cpp:2490]:
```suggestion
				x = Math::sin(u * Math_TAU);
				z = Math::cos(u * Math_TAU);
```

File: scene/resources/3d/primitive_meshes.cpp [hunk 12]
```diff
@@ -2503,8 +2556,12 @@ void TubeTrailMesh::_create_mesh_array(Array &p_arr) const {
 				float r = i;
 				r /= radial_steps;
 
-				float x = sin(r * Math_TAU);
-				float z = cos(r * Math_TAU);
+				float x = 0.0;
+				float z = 1.0;
+				if (i < radial_steps) {
+					x = sin(r * Math_TAU);
+					z = cos(r * Math_TAU);
```
Comment (AThousandShips) [inline scene/resources/3d/primitive_meshes.cpp:2563]:
```suggestion
					x = Math::sin(r * Math_TAU);
					z = Math::cos(r * Math_TAU);
```

File: scene/resources/3d/primitive_meshes.cpp [hunk 13]
```diff
@@ -2568,8 +2625,12 @@ void TubeTrailMesh::_create_mesh_array(Array &p_arr) const {
...
```
Comment (AThousandShips) [inline scene/resources/3d/primitive_meshes.cpp:2632]:
```suggestion
					x = Math::sin(r * Math_TAU);
					z = Math::cos(r * Math_TAU);
```

File: scene/resources/3d/primitive_meshes.cpp [hunk 14]
```diff
@@ -443,15 +443,25 @@ void CapsuleMesh::create_mesh_array(Array &p_arr, const float radius, const floa
...
 
 		v /= (rings + 1);
-		w = sin(0.5 * Math_PI * v);
-		y = radius * cos(0.5 * Math_PI * v);
+		if (j == (rings + 1)) {
+			w = 1.0;
+			y = 0.0;
+		} else {
+			w = sin(0.5 * Math_PI * v);
+			y = cos(0.5 * Math_PI * v) * radius;
```
[the remark this answers is not in the document]
Comment (Lielay9) [reply scene/resources/3d/primitive_meshes.cpp]: Done~ ^^

**Review by clayjohn — APPROVED** [review]

**Comment by Repiteo** [comment]
Thanks!

**Pull request merged**